src/storage/v2/constraints/existence_constraints.cpp

```diff
@@ -55,4 +55,80 @@ void ExistenceConstraints::LoadExistenceConstraints(const std::vector<std::strin
   }
 }
 
+std::variant<ExistenceConstraints::MultipleThreadsConstraintValidation,
+             ExistenceConstraints::SingleThreadConstraintValidation>
+ExistenceConstraints::GetCreationFunction(
+    const std::optional<durability::ParallelizedSchemaCreationInfo> &par_exec_info) {
+  if (par_exec_info.has_value()) {
+    return ExistenceConstraints::MultipleThreadsConstraintValidation{par_exec_info.value()};
+  }
+  return ExistenceConstraints::SingleThreadConstraintValidation{};
+}
+
+[[nodiscard]] std::optional<ConstraintViolation> ExistenceConstraints::ValidateVerticesOnConstraint(
+    utils::SkipList<Vertex>::Accessor vertices, LabelId label, PropertyId property,
+    const std::optional<durability::ParallelizedSchemaCreationInfo> &parallel_exec_info) {
+  auto calling_existence_validation_function = GetCreationFunction(parallel_exec_info);
+  return std::visit(
+      [&vertices, &label, &property](auto &calling_object) { return calling_object(vertices, label, property); },
+      calling_existence_validation_function);
+}
+
+std::optional<ConstraintViolation> ExistenceConstraints::MultipleThreadsConstraintValidation::operator()(
+    utils::SkipList<Vertex>::Accessor &vertices, const LabelId &label, const PropertyId &property) {
+  utils::MemoryTracker::OutOfMemoryExceptionEnabler oom_exception;
+
+  const auto &vertex_batches = parallel_exec_info.first;
+  const auto thread_count = std::min(parallel_exec_info.second, vertex_batches.size());
+
+  MG_ASSERT(!vertex_batches.empty(),
+            "The size of batches should always be greater than zero if you want to use the parallel version of index "
+            "creation!");
+
+  std::atomic<uint64_t> batch_counter = 0;
+  memgraph::utils::Synchronized<std::optional<ConstraintViolation>, utils::SpinLock> maybe_error{};
+  {
+    std::vector<std::jthread> threads;
+    threads.reserve(thread_count);
+
+    auto thread_function = [&label, &property, &vertex_batches, &maybe_error, &batch_counter, &vertices]() {
+      while (!maybe_error.Lock()->has_value()) {
+        const auto batch_index = batch_counter.fetch_add(1, std::memory_order_acquire);
+        if (batch_index >= vertex_batches.size()) {
+          return;
+        }
+        const auto &[gid_start, batch_size] = vertex_batches[batch_index];
+
+        auto vertex_curr = vertices.find(gid_start);
+
+        for (auto i{0U}; i < batch_size; ++i, ++vertex_curr) {
```

Comment:
It must be okay because these batches and everything are created on data recovery from the snapshot. We create each batch to have following data on snapshot creation:
```
// 10) Batch infos
...
//     * vertex batch infos
//        * starting offset of the batch
//        * number of vertices in the batch
```

Otherwise we have bug in snapshot creation.